labgrid/driver/shelldriver.py

```diff
@@ -102,24 +103,38 @@ def run(self, cmd, timeout=30.0):
     @step()
     def _await_login(self):
         """Awaits the login prompt and logs the user in"""
-        self.console.sendline("")
         # TODO use step timeouts
+
+        self.console.sendline("")
+
+        expectations = [self.prompt, self.login_prompt]
+        if self.console_ready != "":
+            expectations.append(self.console_ready)
+
         index, _, _, _ = self.console.expect(
-            [self.prompt, self.login_prompt],
+            expectations,
             timeout=self.login_timeout
         )
+
         if index == 0:
             self.status = 1
             return  # already logged in
-        self.console.sendline(self.username)
-        index, _, _, _ = self.console.expect([self.prompt, "Password: "], timeout=10)
-        if index == 1:
-            if self.password:
-                self.console.sendline(self.password)
-                self.console.expect(self.prompt, timeout=5)
-            else:
-                raise Exception("Password entry needed but no password set")
-        self._check_prompt()
+        elif index == 1:
+            # we need to login
+            self.console.sendline(self.username)
+            index, _, _, _ = self.console.expect([self.prompt, "Password: "], timeout=10)
+            if index == 1:
+                if self.password:
+                    self.console.sendline(self.password)
+                    self.console.expect(self.prompt, timeout=5)
+                else:
+                    raise Exception("Password entry needed but no password set")
+            self._check_prompt()
+        elif index == 2:
+            # we have just activated a console here
+            # lets start over again and see if login or prompt will appear now
+            self.console.sendline("")
+            self._await_login()
```

Comment:
_await_login() will send newline anyway, why do we need two? I'm somewhat skeptical regarding the recursion in this case, an explicit loop over the main expect might be clearer.

Reply:
You are right. _await_login() will send a newline anyway. (Even with the new silence-timeout schema discussed laster.) I'll remove L136.

Regarding the explicit loop:
I am fine with that. I just (mostly unconscious) copied this style from for example the ubootstrategy.py:50.
I will change this too.